integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/OpenLineageRunEventBuilder.java

```diff
@@ -266,7 +292,20 @@ private List<InputDataset> buildInputDatasets(List<Object> nodes) {
                                 .flatMap(Collection::stream)
                                 .map(((Class<InputDataset>) InputDataset.class)::cast))
                     .orElse(Stream.empty()))
-            .collect(Collectors.toList());
+            .collect(
+                Collectors.toMap(
+                    // Key Mapper: Create a DatasetKey from namespace and name
+                    dataset -> new DatasetKey(dataset.getNamespace(), dataset.getName()),
+
+                    // Value Mapper: The dataset itself
+                    Function.identity(),
+
+                    // Merge Function: If duplicates found, keep the first one encountered
+                    (ds1, ds2) -> ds1));
```

Comment:
Would it make sense to use `io.openlineage.client.OpenLineageClientUtils#mergeFacets` instead of picking the first one?

Reply:
wouldn't be easier to use Set instead of Map? You get merge for free